server/src/main/java/org/apache/kafka/server/BrokerFeatures.java

```diff
@@ -0,0 +1,193 @@
+     * supported features.
+     *
+     * @param supportedFeatures The supported features to be compared
+     * @param finalizedFeatures The finalized features to be compared
+     * @return - True if there are any feature incompatibilities found.
+     * - False otherwise.
+     */
+    public static boolean hasIncompatibleFeatures(Features<SupportedVersionRange> supportedFeatures,
+                                                  Map<String, Short> finalizedFeatures) {
+        return !incompatibleFeatures(supportedFeatures, finalizedFeatures, false).isEmpty();
+    }
+
+    /**
+     * Returns the default finalized features that a new Kafka cluster with IBP config >= IBP_2_7_IV0
+     * needs to be bootstrapped with.
+     */
+    public Map<String, Short> defaultFinalizedFeatures() {
+        return supportedFeatures.features().entrySet()
+                .stream()
+                .collect(HashMap::new, (m, e) -> {
+                    String name = e.getKey();
+                    SupportedVersionRange versionRange = e.getValue();
+                    if ("kraft.version".equals(name)) {
+                        m.put(name, (short) 0);
+                    } else {
+                        m.put(name, versionRange.max());
+                    }
+                }, HashMap::putAll);
+    }
+
+    /**
+     * Returns the set of feature names found to be incompatible.
+     * A feature incompatibility is a version mismatch between the latest feature supported by the
+     * Broker, and a provided finalized feature. This can happen because a provided finalized
+     * feature:
+     * 1) Does not exist in the Broker (i.e. it is unknown to the Broker).
+     * [OR]
+     * 2) Exists but the FinalizedVersionRange does not match with the SupportedVersionRange
+     * of the supported feature.
+     *
+     * @param finalized The finalized features against which incompatibilities need to be checked for.
+     * @return The subset of input features which are incompatible. If the returned object
+     * is empty, it means there were no feature incompatibilities found.
+     */
+    public Map<String, Short> incompatibleFeatures(Map<String, Short> finalized) {
+        return BrokerFeatures.incompatibleFeatures(supportedFeatures, finalized, true);
+    }
+
+    public Features<SupportedVersionRange> supportedFeatures() {
+        return supportedFeatures;
+    }
+
+    private static Map<String, Short> incompatibleFeatures(Features<SupportedVersionRange> supportedFeatures,
+                                                           Map<String, Short> finalizedFeatures,
+                                                           boolean logIncompatibilities) {
+        List<IncompatibleFeaturesInfo> incompatibleFeaturesInfo = finalizedFeatures
+                .entrySet()
+                .stream()
+                .map(entry -> transferToIncompatibleFeaturesInfo(supportedFeatures, entry))
+                .filter(info -> info.errorReason != null)
```

Comment:
maybe `transferToIncompatibleFeaturesInfo` should return the feature with error to avoid extra filter?